### PR TITLE
Feature: Improve Dio exception reports

### LINKED ISSUE #715: [dio] Improve fingerprinting
I tested the new `sentry_dio` package a little bit.

My issue is that due to how DIO throws `DioError`s, the stacktrace is always the same (that's a problem in itself and probably needs to be improved) - this results in all captured errors having the same fingerprint.

Additionally a `DioError` can wrap a `cause` error which may be the real reason why the request failed, this currently does not get reported.

@ueman What are you thoughts on this?

### PR DESCRIPTION
**Pull request by ueman**
## :scroll: Description
Adds an event processor, which add chained exceptions for the inner exception of `DioError`, 
see https://sentry.io/share/issue/638a3758a43d4f61bc2a0c62acdefd69/

This one should be merged after https://github.com/getsentry/sentry-dart/pull/728

## :bulb: Motivation and Context

Fixes https://github.com/getsentry/sentry-dart/issues/715

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated the docs if needed
- [x] All tests passing
- [ ] No breaking changes

### CODE REVIEW

File: flutter/example/lib/main.dart
Comment (ueman) [inline flutter/example/lib/main.dart]: After options is exposed on the hub, I can move this to rest of the Dio setup, so that it's still only one step to configure.

**Comment by kuhnroyal** [comment]
I never get any `DioError` in the `FailedRequestClient` , the whole block here: https://github.com/flutterchina/dio/blob/develop/dio/lib/src/dio_mixin.dart#L620. is executed later.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [inline dio/lib/src/dio_event_processor.dart]: Both of these aren't exposed, should I expose them or ignore the `implementation_imports` warning?

**Comment by kuhnroyal** [comment]
This works, I have the correct in-app grouping.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_event_processor.dart]: This can be `null`, not sure what happens then.
Also could wrap it in a `ThrowableMechanism` similar to the `FailedRequestClient`.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_event_processor.dart]: `...?` not supported in min Dart version?

**Review by kuhnroyal — COMMENTED** [review]
I basically copied this and and added some more code from `FailedRequestClient` in order to add a `SentryRequest` to the event. And I disabled the `FailedRequestClient`. This combination now gives me a pretty good result.
The only problem I see, is that the error is not tied to the transaction in any way.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Not sure, I'll check that.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Yeah, I know. But this is not yet tested at all, that's why it's a draft :D

**Comment by ueman** [comment]
> I basically copied this and and added some more code from `FailedRequestClient` in order to add a `SentryRequest` to the event. And I disabled the `FailedRequestClient`. This combination now gives me a pretty good result.

That's actually a pretty good idea.

> The only problem I see, is that the error is not tied to the transaction in any way. 

As this integration doesn't start any transactions, you basically have to write code like the one below anyway
```dart
  final transaction = Sentry.getSpan() ??
      Sentry.startTransaction(
        'dio-web-request',
        'request',
        bindToScope: true,
      );
  try {
    final response = await dio.get<String>('https://flutter.dev/');
    transaction.spanStatus = SpanStatus.ok();
  } catch (exception, stackTrace) {
    transaction.throwable = exception;
    transaction.spanStatus = SpanStatus.internalError();
    Sentry.captureException(exception, stacktrace);
  } finally {
    await transaction.finish();
  }
```
and then it is tied together.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [inline dio/lib/src/dio_event_processor.dart]: After doing this, I would like to set `dioError.stackTrace` (and maybe `dioError.error`) to `null`, so the report on Sentry looks much nicer.
Also, because we're adding an extra SentryException for those, it should be okay. Any opinions on that?

**Comment by kuhnroyal** [comment]
> As this integration doesn't start any transactions, you basically have to write code like the one below anyway

True, it is all tied to my Flutter route transactions.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart]: Not sure, no opinion yet.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_event_processor.dart]: I found that setting the description screws up the UI in Sentry, the mechanism name is no longer showing. There is no description set for `runZonedGuarded` etc.

File: dart/lib/src/http_client/sentry_http_client.dart
Comment (kuhnroyal) [inline dart/lib/src/http_client/sentry_http_client.dart]: Somewhat unrelated, but can we somehow use the `ValidateStatus` from the Dio options as default for `failedRequestStatusCodes`?

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_event_processor.dart]: Hmm I don't see any difference.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_event_processor.dart]:
```suggestion
class DioEventProcessor implements EventProcessor {
```

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Yeah, I also noticed it. It could be because the original exception was already converted to a SentryException at this point. I'll have to look into it some more.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: But it should still show the original Mechanism for the DioError. 

@marandaneto How do order the exceptions, so that the DioError is the main exception and the the inner exception is the second?

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Does that make any difference here?

File: dart/lib/src/http_client/sentry_http_client.dart
Comment (ueman) [reply dart/lib/src/http_client/sentry_http_client.dart]: I'm thinking of removing the `FailingRequestAdapter` all together because of this. If Dio throws on specific status codes anyway, there's no need for that functionality.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart]: Well it is an `interface`, but I am coming from the Java world so when I see `extends`, I expect some implementation to exist in the superclass. Not really important, up to you :)

**Comment by kuhnroyal** [comment]
When a description is set on the mechanism, the type is not shown in Sentry UI. Not sure why.

Bildschirmfoto 2022-01-27 um 10 56 09
Bildschirmfoto 2022-01-27 um 10 56 18

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart]: It always shows `runZonedGuarded` for the `DioError`.
But not for the cause when the description is set, see https://github.com/getsentry/sentry-dart/pull/718#issuecomment-1023036545

Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart]: Might be something buggy in the UI.

**Comment by ueman** [comment]
> When a description is set on the mechanism, the type is not shown in Sentry UI. Not sure why.
> 
> Bildschirmfoto 2022-01-27 um 10 56 09 Bildschirmfoto 2022-01-27 um 10 56 18

Yeah, that's also true for `FlutterError` which sometime have the description set. 
I don't know whether that's working as expected or not. Maybe @marandaneto know. 
Probably this is worth an issue at the Sentry repo itself.

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart]: They are internal, what we can do is to add them to `SentryOptions` and use the getter from there, so we have also a single instance of it.

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart]: the sorting matter, See `SentryStackTraceFactory#getStackFrames`, we `reversed` the list because of that.
`DioError` should be the last item on the exception list.

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [inline dio/lib/src/dio_event_processor.dart]: You don't need a mechanism for the inner exception, only for the "main" one.

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [inline dio/lib/src/dio_event_processor.dart]: What if we always use the `dioError.error` for the `dio` integration, isn't that any better? is there info on `dioError` that's not on `dioError.error`?

**Comment by marandaneto** [comment]
> > When a description is set on the mechanism, the type is not shown in Sentry UI. Not sure why.
> > Bildschirmfoto 2022-01-27 um 10 56 09 Bildschirmfoto 2022-01-27 um 10 56 18
> 
> Yeah, that's also true for `FlutterError` which sometime have the description set. I don't know whether that's working as expected or not. Maybe @marandaneto know. Probably this is worth an issue at the Sentry repo itself.

I agree, an issue there would be better, I don't know the answer either.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Yep, there's info on `DioError` which is not available on `DioError.error`. Also, `DioError.error` is not always available.

Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart]: Makes sense to report both whenever possible then.

I'm wondering if we can make it without the EventProcessor.
What if we pass a List of exceptions to `event.throwable` since its a `dynamic` field and in the `sentryExceptionFactory` we special case if it's a `List`, we traverse the list converting to `SentryException`, didn't try, but it could work.

The downside is if somebody does `throw randomList` and this List isn't this special list of exceptions, but rather an App's list, we need to consider that too.

Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart]: We don't have the `DioError` available earlier. The `SentryDioClientAdapter` is not aware of these errors. The `DioError`s usually are manually caught and captured or come from the zones uncaught handler.

Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: But then we would have to know how to create list from `DioError` inside the `sentry` sdk. But that doesn't know about Dio.

Is there a way to register a `SentryExceptionFactory` for specific exceptions?

Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart]: I see, then an event processor makes sense, we just need to add it automatically, so fewer steps for the final user.
A thing to consider is, an App may have more than one `dio` instance, so when `addSentry` is called, we should not register the event processor twice.

Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: > I see, then an event processor makes sense, we just need to add it automatically, so fewer steps for the final user.

See https://github.com/getsentry/sentry-dart/pull/718#discussion_r792821809

> A thing to consider is, an App may have more than one dio instance, so when addSentry is called, we should not register the event processor twice.

Good point.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: I would like to add some explanation to the inner exception to describe that it is an inner exception. How do I do that without a mechanism?

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_event_processor.dart]: I tested a lot today, there is even more to this I think.
When `dioError.error is Error` then we have to use  `dioError.error.stackTrace`. We may even have to add 2 chained exceptions here.
For example some form of serialization error or `SocketException`s have a distinct stackTrace in `dioError.error.stackTrace` but the caller information from your own code is in `dioError.stackTrace`. 
On the other hand for HTTP response errors, `dioError.error` is just a String so there is no additional stackTrace.

So there can actually be 3 separate stackTraces for each DioError.
1. The on you get with `catch(e, stackTrace)`- this one is mostly useless as it always points to some internal Dio code. But it is likely the one being captured with `captureException` by the user. This is the one you are removing now. (mostly useless)
2. The `dioError.stackTrace` - this one is generated with `StackTrace.current` inside Dio and most of the time contains the correct call stack from the users code (inApp). (important)
3. If `dioError.error is Error` then there is a `dioError.error.stackTrace` - this one contains detailed information about potential inner errors which happen during serialization/send/receive. (kinda important)

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [inline dio/lib/src/dio_event_processor.dart:56]: I think we can just drop these exceptions?

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Thank you for this. That's a great deep dive.
Also something to think about in https://github.com/flutterchina/dio/pull/1405

I wonder if this also applies to pure Dart code. Errors (in contrast to Exceptions) always have their own stacktrace. So if you try-catch an error, you have two stacktraces. Errors are kinda defined as mistakes a developer makes at the time of writing but then again people probably still try-catch errors 🤷‍♂️

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart:56]: What makes you think so?

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart]: Dart 2.16 added `Error.throwWithStackTrace(error, stackTrace)` which should probably be used in Dio 5.0.
This would solve some of the problems.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart:56]: Well for me, this is always just the `DioError` with the wrong stacktrace making it completely useless.
And we are adding the correct `DioError`.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Yeah, or just to completely remove any catching of errors/excetptions in Dio itself

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart:56]: Yeah, for me that's also the case. But there may be other event processers at work which add to it, and we shouldn't drop those. We should only replace the DioError and keep everything as is.

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart]: Everything but the main exception is an inner exception (or chained exception), even the UI shows them below the main exception, why do you need a mechanism for that? just trying to understand the use case.

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart:56]: Yea I agree

File: dio/lib/src/dio_event_processor.dart
Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart]: But then they can't be wrapped into a `DioError` :)

Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: Which isn't a problem if it doesn't add anything valuable to the original exception and you don't run into problems with masking or losing the original exception/stacktrace.

File: dart/lib/src/http_client/sentry_http_client.dart
Comment (ueman) [reply dart/lib/src/http_client/sentry_http_client.dart]: https://github.com/getsentry/sentry-dart/pull/728

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart]: https://github.com/getsentry/sentry-dart/pull/727

**Comment by marandaneto** [comment]
There are a lot of trade-offs here and I'd need some time to review this and decide how we could improve the chained exceptions and grouping for `dio`, the discussions are insightful tho, thank you @ueman and @kuhnroyal

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [inline dio/lib/src/dio_event_processor.dart:116]: You're using request props configuration for the `response` object, which is wrong.
Looks like `DioError` does not have access to the `request` data, but only `response`.

Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart:116]: Btw adding response data requires discussion yet, See https://github.com/getsentry/sentry-dart/issues/624#issuecomment-1012296140 due to PII concerns, etc...

Comment (kuhnroyal) [reply dio/lib/src/dio_event_processor.dart:116]: Access to the request options is available.

Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart:116]: Oh looks like it's a different name, `error.requestOptions.data` and friends would work, indeed.

**Comment by marandaneto** [comment]
@ueman and @kuhnroyal when it throws e.g. `SocketException`, the stackTrace has only one frame and it is low level, those events are grouped together, I don't see a big problem for that since it does not matter where it comes from, there's no network or you can't do much about it anyway.

If it's about `failedRequestStatusCodes`, it uses the `StackTrace.current` and it does not contain the App's frames, fair, events may be grouped together since the frames are almost always the same, but the important bit here is, which URL is throwing? and this is answered by the events, does it matter if the same API is called from different places?

`DioError` seems to be richer in both cases.
If you have access to the `DioError` and it throws because of e.g. 404, you have access to the stack trace with App's frames.
If you have access to the `DioError` and it throws because of e.g. no network, you have access to `dio.error` which is the wrapped `SocketException`, also `dio.stackTrace` has Apps frames.

so `DioError` is always better, `DioError.stackTrace` is always better than the captured `stackTrace` via the `catch block`, `dio.error` could be a chained exception.

Do you know if using Dio Interceptors, do we have access to the DioError?
What was the trade-off between the Adapter vs the Interceptor?

**Comment by kuhnroyal** [comment]
> so `DioError` is always better, `DioError.stackTrace` is always better than the captured `stackTrace` via the `catch block`, `dio.error` could be a chained exception.

Yes thats right.

Adapter was easier to implement and may be easier to maintain, as the implementation is similar to the one for the `http` package.
But an adapter doesn't know about the `DioError`, an interceptor does.
The interceptor is harder to configure correctly, as interceptors actually can prevent following interceptors from running or modify the response/error.

**Comment by marandaneto** [comment]
> > so `DioError` is always better, `DioError.stackTrace` is always better than the captured `stackTrace` via the `catch block`, `dio.error` could be a chained exception.
> 
> Yes thats right.
> 
> Adapter was easier to implement and may be easier to maintain, as the implementation is similar to the one for the `http` package. But an adapter doesn't know about the `DioError`, an interceptor does. The interceptor is harder to configure correctly, as interceptors actually can prevent following interceptors from running or modify the response/error.

Indeed, we do something similar for Java-OkHttp and it was quite straightforward -> https://github.com/getsentry/sentry-java/blob/main/sentry-android-okhttp/src/main/java/io/sentry/android/okhttp/SentryOkHttpInterceptor.kt
I don't know the internals of `dio` or how it handles interceptors, and errors, but it should not be that different.

We could try since I don't see any other way to improve the current impl. without `DioError`.

@ueman solution with the EventProcessor also seems to be a good idea, I'm wondering which one is better.

**Comment by ueman** [comment]
> The interceptor is harder to configure correctly, as interceptors actually can prevent following interceptors from running or modify the response/error.

That's the point I'm concerned about.

Depending on the configuration we might report errors even if they're later discarded and we could also get the data wrong as it might be modified after recording. Also we might not know about errors because a potential SentryDioInterceptor wasn't even called. 

I'm not quite sure how hard or error prone it is to get it right, so I'm gladly hear your thoughts on it.

We probably shouldn't promote the dio integration to stable until this discussion is solved tho 😄

**Comment by kuhnroyal** [comment]
I have used an interceptor based approach so far and it worked pretty well, but the configuration problems are a point of concern.

Maybe it just needs to be documented, I mean now we have documented that `..configureSentry()` needs to be called last. With an interceptor we just leave that and default just add the Sentry interceptor it at index 0? Dunno.

On the other hand this "feature" may be what we want. Let's says you have an interceptor which handles OAuth refreshes when there is a 401/whatever response. I can then choose to not forward this response error because I am instead refreshing the token and re-execute the original request. I don't want to see this in Sentry.

**Comment by marandaneto** [comment]
If we don't capture events because any other interceptor has discarded it, well, it really depends on the setup and what the other interceptors do, it's a best-effort approach, if the data comes in handy to our interceptor, we're able to enrich and capture errors, otherwise not.

For 401 errors, you can either disable not passing to the error code range or even use the `beforeSend` callback to drop it.

The interceptor should be configurable enough and fit most of the "normal" use cases, users are free to either configure accordingly or just not use it.

I don't think there's a solution for every edge case, but it should be possible for the majority that does simple REST calls, etc...

What do you think?

**Comment by ueman** [comment]
> What do you think?

For performance tracing it would be very bad to have a span/transaction which is not closed. So performance tracing is going to stay as is. 
The FailedRequestAdapter should be converted to an interceptor. The DioEventProcessor is still useful for chained stack traces, so it's also going to stay.

Did I get that right?

**Comment by kuhnroyal** [comment]
> For performance tracing it would be very bad to have a span/transaction which is not closed. So performance tracing is going to stay as is.

Is it that bad? Or is it the responsibility of the developer to adapt his code?

> The DioEventProcessor is still useful for chained stack traces, so it's also going to stay.

We can get chained exceptions from interceptor as well but then it doesn't work for try/catch... And do you need to capture with try/catch when it was already captured in an interceptor?


How does it work with the `http` package? Uncaught vs adapter/client?

**Comment by ueman** [comment]
> How does it work with the `http` package? Uncaught vs adapter/client?

`http` only throws on network failures and it has no interceptors. So everything is done as adapter/client.

**Comment by kuhnroyal** [comment]
We will end. up with duplicate events If we capture all errors in an interceptor and then automatically have the same errors captured in the zone...

**Comment by kuhnroyal** [comment]
@ueman I didn't get around to testing again yet, but have a slightly related question.

What do you think about adding a way to pass a custom parent span to a request via Dio's `extra` map, that the `TracingClientAdapter` will use by default and fallback to `Sentry.getSpan()` if it is not set?

Like so:
```dart
  @override
  Future<ResponseBody> fetch(
    RequestOptions options,
    Stream<Uint8List>? requestStream,
    Future? cancelFuture,
  ) async {
    // see https://develop.sentry.dev/sdk/performance/#header-sentry-trace
    final currentSpan = options.extra['sentrySpan'] is SentrySpan ? options.extra['sentrySpan']! as SentrySpan : _hub.getSpan();
    final span = currentSpan?.startChild(
      'http.client',
      description: '${options.method} ${options.uri}',
    );
```

**Comment by ueman** [comment]
> @ueman I didn't get around to testing again yet, but have a slightly related question.

Nothing much has changed regarding the functionality of this PR. Mostly clean up. There's another one in https://github.com/getsentry/sentry-dart/pull/728. Regarding https://github.com/getsentry/sentry-dart/pull/718#discussion_r794086467, I'll raise another PR which should enable that for all errors, even non DioErrors.

> What do you think about adding a way to pass a custom parent span to a request via Dio's `extra` map, that the `TracingClientAdapter` will use by default and fallback to `Sentry.getSpan()` if it is not set?

That should be doable, but better in another PR 😅

Edit: https://github.com/getsentry/sentry-dart/issues/751

File: dio/test/sentry_dio_extension_test.dart
Comment (marandaneto) [inline dio/test/sentry_dio_extension_test.dart]: We could make a Fixture here.

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [inline dio/lib/src/dio_event_processor.dart]: What's the reason for not using it's type? `String` in this case.

**Comment by marandaneto** [comment]
@ueman can you share a link of an issue after those changes or a screenshot? thanks.
Looks like this PR indeed makes it much better.
Logic wise LGTM, @brustolin please review :)

**Comment by ueman** [comment]
`await dio.get<String>('https://flutter.dev/asdfasg');`, throws because 404 status code: https://sentry.io/organizations/sentry-sdks/issues/3007144862/events/31495690091f4d7684c339e6fd9a6ebc/?project=5428562

`await dio.post<String>('https://flutter.dev/', data: Exception());`, throws because `Exception` can't be serialized: https://sentry.io/organizations/sentry-sdks/issues/3009008663/events/fcb1d27e11be4af4b71b25b6c04128c6/?project=5428562

**Comment by brustolin** [comment]
@marandaneto, as you said, code looks good. 
But I don't really know how EventProcessor works in the Dart SDK, need to get familiar with this.

**Comment by ueman** [comment]
> @marandaneto, as you said, code looks good. But I don't really know how EventProcessor works in the Dart SDK, need to get familiar with this.

Feel free to ask (here or on Discord), I'll probably can answer at least some questions.

**Comment by marandaneto** [comment]
@ueman sentry-dio / package-analysis is unhappy, otherwise LGTM.

**Review by marandaneto — APPROVED** [review]
@ueman thanks for making the `dio` integration better :)

**Comment by marandaneto** [comment]
@kuhnroyal also thanks for the valuable tests and insights.

**Comment by ueman** [comment]
I would guess it's because pana doesn't check the dependency overrides? The getter which pana complains about isn't available in the latest stable but on main. I don't think I can do anything about that.

Pana isn't happy because of:

------

### [x] 0/30 points: code has no errors, warnings, lints, or formatting issues

Found 3 issues. Showing the first 2:<details>
<summary>
ERROR: The getter 'exceptionFactory' isn't defined for the type 'SentryOptions'.
</summary>`lib/src/dio_event_processor.dart:20:16````
   ╷
20 │       _options.exceptionFactory;
   │                ^^^^^^^^^^^^^^^^
   ╵
```
To reproduce make sure you are using the [lints_core](https://pub.dev/packages/lints) and run `dart analyze lib/src/dio_event_processor.dart`
</details>
<details>
<summary>
ERROR: The method 'shouldAddBody' isn't defined for the type 'MaxRequestBodySize'.
</summary>
`lib/src/dio_event_processor.dart:127:31`
```
    ╷
127 │       if (_maxRequestBodySize.shouldAddBody(data.codeUnits.length)) {
    │                               ^^^^^^^^^^^^^
    ╵
```
To reproduce make sure you are using the [lints_core](https://pub.dev/packages/lints) and run `dart analyze lib/src/dio_event_processor.dart`
</details>

------

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [inline dio/lib/src/dio_event_processor.dart:5]: @ueman is that because `sentry_exception_factory` isn't exported?

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [inline dio/lib/src/dio_event_processor.dart:20]: Maybe if we use `_options.exceptionFactory` directly, we avoid this linter ignore.

File: dio/lib/src/sentry_dio_extension.dart
Comment (marandaneto) [inline dio/lib/src/sentry_dio_extension.dart:19]: The reason for that is because `dio` itself already throws in some cases, right?

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart:5]: I don't think so. The error message says that the getter do not exist, which is true for 6.3.0 but at the current tip of main it exists. Also, locally I don't get any errors/warnings. Only if I run it against the current stable I get the same warnings.

File: dio/lib/src/sentry_dio_extension.dart
Comment (ueman) [reply dio/lib/src/sentry_dio_extension.dart:19]: Yes, exactly

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart:5]: Adding `export 'src/sentry_exception_factory.dart';` to `sentry.dart` fixes the issue, since its a property on SentryOptions, its fine to be there too (Tested it locally).

File: dio/lib/src/dio_event_processor.dart
Comment (marandaneto) [reply dio/lib/src/dio_event_processor.dart:20]: Nvm, the problem would still exist, Since its `@internal`, the linter only likes it if used within its own package.

**Comment by marandaneto** [comment]
> I would guess it's because pana doesn't check the dependency overrides? The getter which pana complains about isn't available in the latest stable but on main. I don't think I can do anything about that.
> 
> Pana isn't happy because of:
> 
> ### [x] 0/30 points: code has no errors, warnings, lints, or formatting issues
> Found 3 issues. Showing the first 2:
> 
> ERROR: The getter 'exceptionFactory' isn't defined for the type 'SentryOptions'.
> ERROR: The method 'shouldAddBody' isn't defined for the type 'MaxRequestBodySize'.

Indeed, I believe `pana` should respect `dependency_overrides` or at least offer an option to respect it since multi-module projects would run into this issue very often.
Maybe worth filing an issue on https://github.com/dart-lang/pana/ as a feature request.
Let's merge and not block on this then :)

**Comment by marandaneto** [comment]
I'm fine if you want to address https://github.com/getsentry/sentry-dart/pull/718#discussion_r807145080 or keep as it is, your call.
I'm good to merge as it is.

File: dio/lib/src/dio_event_processor.dart
Comment (ueman) [reply dio/lib/src/dio_event_processor.dart:5]: I would say we keep it as is.

**Pull request merged**
